ibtk/src/lagrangian/FEDataManager.cpp

```diff
@@ -620,6 +620,119 @@ FEDataManager::reinitElementMappings()
+        const int rank = IBTK_MPI::getRank();
+        const int n_procs = IBTK_MPI::getNodes();
+        const MeshBase& mesh = getEquationSystems()->get_mesh();
+        std::unique_ptr<PetscVector<double> > X_petsc_vec = buildIBGhostedVector(COORDINATES_SYSTEM_NAME);
+        *X_petsc_vec = *getCoordsVector();
+        X_petsc_vec->close();
+        const double* const X_local_soln = X_petsc_vec->get_array_read();
+        const DofMap& X_dof_map = d_fe_data->d_es->get_system(COORDINATES_SYSTEM_NAME).get_dof_map();
+
+        std::vector<int> node_ranks(mesh.parallel_n_nodes());
+        Pointer<PatchLevel<NDIM> > level = d_hierarchy->getPatchLevel(d_fe_data->d_level_number);
+        int local_patch_num = 0;
+        std::vector<dof_id_type> X_idxs;
+        for (PatchLevel<NDIM>::Iterator p(level); p; p++, ++local_patch_num)
+        {
+            const Pointer<Patch<NDIM> > patch = level->getPatch(p());
+            const Pointer<CartesianPatchGeometry<NDIM> > patch_geom = patch->getPatchGeometry();
+            const double* const patch_x_lower = patch_geom->getXLower();
+            const double* const patch_x_upper = patch_geom->getXUpper();
+
+            for (const Node* n : d_active_patch_node_map[local_patch_num])
+            {
+                IBTK::Point X;
+                bool inside_patch = true;
+                for (unsigned int d = 0; d < NDIM; ++d)
+                {
+                    IBTK::get_nodal_dof_indices(X_dof_map, n, d, X_idxs);
+                    X[d] = X_local_soln[X_petsc_vec->map_global_to_local_index(X_idxs[0])];
+                    // Due to how SAMRAI computes patch boundaries, even if the
+                    // patch's domain is [0, 1]^2 the patches on the boundary
+                    // may not actually end at 1.0. Hence allow a small
+                    // tolerance here to account for the case where two patches
+                    // are adjacent to each-other but their patch boundaries
+                    // don't quite line up:
+                    const double x_lower = patch_x_lower[d] - std::max(1.0, std::abs(patch_x_lower[d])) * 1e-14;
+                    const double x_upper = patch_x_upper[d] + std::max(1.0, std::abs(patch_x_upper[d])) * 1e-14;
+                    inside_patch = inside_patch && (x_lower <= X[d] && X[d] <= x_upper);
+                }
+                if (inside_patch)
+                {
+                    node_ranks[n->id()] = rank + 1;
+                }
+                else
+                {
+                    // Points are allowed to be outside the domain - they simply
+                    // are no longer used for IB calculations.
+                    bool inside_hier = true;
+                    for (unsigned int d = 0; d < NDIM; ++d)
+                    {
+                        // Like above - permit points very close to the boundary
+                        // to pass the check and treat them as being outside
```

Comment:
Do we want to treat them as "outside" or "inside" if they are just at the boundary? I would vote for "inside".

Reply:
Here I mean physical boundary --- apologies if I am misinterpreting.

Reply:
Here points very close to the boundary of the domain (the bounding box on the patch hierarchy itself) are treated as being outside the domain. The logic is complicated - if a Node is inside, or nearly inside, a patch then it passes the check. If a Node is *outside* or *nearly outside* the domain then it also passes the check.